### PR TITLE
Reduce order contents complexity

### PR DESCRIPTION
**Pull request by mbj**
Extraction from #177

### CODE REVIEW

**Comment by mbj** [comment]
@dkubb This is ready for review (and as time of comment rebased on HEAD).

File: api/app/controllers/spree/api/line_items_controller.rb [hunk 1]
```diff
@@ -1,16 +1,12 @@
 module Spree
   module Api
     class LineItemsController < Spree::Api::BaseController
-      class_attribute :line_item_options
-
-      self.line_item_options = []
-
       def create
-        variant = Spree::Variant.find(params.fetch(:line_item)[:variant_id])
+        variant = Spree::Variant.find(params.fetch(:line_item).fetch(:variant_id))
```
Comment (dkubb) [inline api/app/controllers/spree/api/line_items_controller.rb]: This method calls `params.fetch(:line_item)` twice. I would suggest that it assign it to a variable and use it in both places.

Comment (mbj) [inline api/app/controllers/spree/api/line_items_controller.rb]: Good catch, there is actually a private currently unused `line_item_params` method I can hook.

File: api/app/controllers/spree/api/line_items_controller.rb [hunk 2]
```diff
@@ -50,18 +46,18 @@ def find_line_item
...
         def line_items_attributes
-          {line_items_attributes: {
-              id: params[:id],
-              quantity: params[:line_item][:quantity],
-              options: line_item_params[:options] || {}
-          }}
+          {
+            line_items_attributes: {
+              id:       params.fetch(:id),
+              quantity: params.fetch(:line_item).fetch(:quantity)
```
Comment (dkubb) [inline api/app/controllers/spree/api/line_items_controller.rb]: This should probably use `line_item_params` instead of `params` explicitly. When possible nothing should reference `params` directly except for the method that specifies which attributes are allowed.

Comment (mbj) [inline api/app/controllers/spree/api/line_items_controller.rb]: Too bad there are no type constraints. Here also we maybe later want to specify a params schema instead relying on unexplicit coercions of AR or other layers.

File: core/app/models/spree/order_contents.rb [hunk 3]
```diff
@@ -7,39 +7,36 @@ def initialize(order)
     end
 
     def add(variant, quantity = 1, options = {})
-      line_item = add_to_line_item(variant, quantity, options)
-      after_add_or_remove(line_item, options)
+      target_shipment = options[:shipment]
```
Comment (dkubb) [inline core/app/models/spree/order_contents.rb]: Is the `:shipment` key required? My first thought was to suggest using `Hash#fetch` here, but why not change the method signature to make it required, eg:

```ruby
def add(variant, quantity = 1, shipment:)
  line_item = add_to_line_item(variant, quantity, shipment)
  after_add_or_remove(shipment)
  line_item
end
```

Or even:

```ruby
def add(variant, quantity = 1, shipment:)
  add_to_line_item(variant, quantity, shipment).tap do
    after_add_or_remove(shipment)
  end
end
```

File: core/app/models/spree/order_contents.rb [hunk 4]
```diff
@@ -7,39 +7,36 @@ def initialize(order)
...
+      line_item = add_to_line_item(variant, quantity, target_shipment)
+      after_add_or_remove(target_shipment)
+      line_item
     end
 
     def remove(variant, quantity = 1, options = {})
```
Comment (dkubb) [inline core/app/models/spree/order_contents.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/181/files#r58942500

File: core/app/models/spree/order_contents.rb (hunk 3 above)
Comment (mbj) [inline core/app/models/spree/order_contents.rb]: Because the shipment is not required. And I nuke the entire argument in the next PR.

Comment (mbj) [inline core/app/models/spree/order_contents.rb]: Also quantity will be required than. There are only 2 of 10 call sides from the domain that actually provide a shipment and here its not actually needed, as the shipment can be set at this call sides also.

Comment (dkubb) [inline core/app/models/spree/order_contents.rb]: > Because the shipment is not required.

Ahh, then you could just have the method signature be:

```def add(variant, quantity = 1, shipment: nil)```

> And I nuke the entire argument in the next PR.

Ahh ok, I wasn't looking further ahead, just looking at the code in front of me for the review.

If it were me I'd probably still make the change so that each step in the transformation is locally optimal in comparison to the original code.

File: core/app/models/spree/order_contents.rb [hunk 5]
```diff
@@ -7,39 +7,36 @@ def initialize(order)
...
     private
-      def after_add_or_remove(line_item, options = {})
+
+      def after_add_or_remove(target_shipment)
         persist_totals
-        shipment = options[:shipment]
-        shipment.present? ? shipment.update_amounts : order.ensure_updated_shipments
-        PromotionHandler::Cart.new(order, line_item).activate
+        target_shipment ? target_shipment.update_amounts : order.ensure_updated_shipments
+        PromotionHandler::Cart.new(order).activate
```
Comment (dkubb) [inline core/app/models/spree/order_contents.rb]: It's not an issue that the order of this line and the line above are swapped from the original in `#update_cart` ?

File: core/app/models/spree/order_contents.rb (hunk 3 above)
Comment (mbj) [inline core/app/models/spree/order_contents.rb]: I like the proposal for a local optimium increase.

File: core/app/models/spree/order_contents.rb [hunk 6]
```diff
@@ -65,33 +62,29 @@ def persist_totals
...
         line_item.price           = variant.price_in(order.currency).amount
 
         line_item.save!
         line_item
       end
 
-      def remove_from_line_item(variant, quantity, options = {})
-        line_item = grab_line_item_by_variant(variant, true, options)
+      def remove_from_line_item(variant, quantity, target_shipment)
+        line_item = order.find_line_item_by_variant(variant) or fail ActiveRecord::RecordNotFound, "Line item not found for variant: #{variant.sku}"
```
Comment (dkubb) [inline core/app/models/spree/order_contents.rb]: I know we probably don't have any line limits in place for rubocop, and I think we should, but in the meantime I would probably suggest wrapping this overly long line at 79 chars.

File: core/app/models/spree/order_contents.rb (hunk 5 above)
Comment (mbj) [inline core/app/models/spree/order_contents.rb]: It does not violate specs. Nor I'm aware of any cart promotion that depends on the shipments right now. Deduplicating the logic is hence more important to me.

File: core/app/models/spree/promotion_handler/cart.rb [hunk 7]
```diff
@@ -13,17 +13,12 @@ module PromotionHandler
...
-      def initialize(order, line_item=nil)
-        @order, @line_item = order, line_item
-      end
+      include Concord.new(:order)
 
       def activate
         promotions.each do |promotion|
-          if (line_item && promotion.eligible?(line_item)) || promotion.eligible?(order)
-            promotion.activate(line_item: line_item, order: order)
+          if promotion.eligible?(order)
```
Comment (dkubb) [inline core/app/models/spree/promotion_handler/cart.rb]: There's no need to check for the `line_item` being defined? Have you tested running the tests with a guard clause that raises if the `line_item` passed to the constructor is nil?

Comment (mbj) [inline core/app/models/spree/promotion_handler/cart.rb]: I removed the only call side that passes in the line item. And changed the arity of `PromotionHandler::Cart` from 2 to 1 only accepting `order`. This is an implicit proof no other call side provides the line item.

File: core/app/models/spree/order_contents.rb [hunk 8]
```diff
@@ -6,40 +6,35 @@ def initialize(order)
       @order = order
     end
 
-    def add(variant, quantity = 1, options = {})
-      line_item = add_to_line_item(variant, quantity, options)
-      after_add_or_remove(line_item, options)
+    def add(variant, quantity = 1, shipment: nil)
+      line_item = add_to_line_item(variant, quantity, shipment)
```
Comment (dkubb) [inline core/app/models/spree/order_contents.rb]: What about using `Object#tap` here to simplify the method body ast to:

```ruby
add_to_line_item(variant, quantity, shipment).tap do
  after_add_or_remove(shipment)
end
```

File: core/app/models/spree/order_contents.rb [hunk 9]
```diff
@@ -6,40 +6,35 @@ def initialize(order)
...
+      after_add_or_remove(shipment)
+      line_item
     end
 
-    def remove(variant, quantity = 1, options = {})
-      line_item = remove_from_line_item(variant, quantity, options)
-      after_add_or_remove(line_item, options)
+    def remove(variant, quantity = 1, shipment: nil)
+      line_item = remove_from_line_item(variant, quantity, shipment)
```
Comment (dkubb) [inline core/app/models/spree/order_contents.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/181/files#r58950519

File: core/app/models/spree/order_contents.rb [hunk 10]
```diff
@@ -65,19 +60,15 @@ def persist_totals
         order_updater.update
       end
 
-      def add_to_line_item(variant, quantity, options)
-        fail 'redundant :currency option' if options.key?(:currency)
-        shipment = options[:shipment]
-        line_item = grab_line_item_by_variant(variant)
+      def add_to_line_item(variant, quantity, shipment)
+        line_item = order.find_line_item_by_variant(variant)
```
Comment (dkubb) [inline core/app/models/spree/order_contents.rb]: You should probably remove one more blank line after this.

File: core/app/models/spree/order_contents.rb [hunk 11]
```diff
@@ -88,10 +79,13 @@ def add_to_line_item(variant, quantity, options)
...
+      def remove_from_line_item(variant, quantity, shipment)
+        line_item = order.find_line_item_by_variant(variant) or
+          fail ActiveRecord::RecordNotFound, "Line item not found for variant: #{variant.sku}"
+
         line_item.quantity -= quantity
-        line_item.target_shipment= options[:shipment]
+
+        line_item.target_shipment = shipment
 
         if line_item.quantity.zero?
```
Comment (dkubb) [inline core/app/models/spree/order_contents.rb]: I realize this isn't in scope, but if the `quantity` is greater than `line_item.quantity` will it mean this conditional will not return true? If so, maybe this should be changed to: `unless line_item.quantity > 0`

File: core/app/models/spree/promotion_handler/cart.rb (hunk 7 above)
Comment (dkubb) [inline core/app/models/spree/promotion_handler/cart.rb]: > changed the arity of PromotionHandler::Cart.new from 2 to 1 only accepting order. This is an implicit proof no other call side provides the line item.

Right, that's good. I am happy with this change.

**Comment by dkubb** [comment]
@mbj I finishes a pass over this PR.

File: core/app/models/spree/order_contents.rb (hunk 11 above)
Comment (mbj) [inline core/app/models/spree/order_contents.rb]: I thought about the same issue, but than realized public API `#remove` is only called from internal call sides that do know the current quantity. So these should never be able to call `quantity > line_item.quantity` case, and if that happens regardless the domain validation error on `LineItem#quantity` will catch it.

**Comment by mbj** [comment]
@dkubb This is ready for the next pass.
